### PR TITLE
Adds kw `remove_destination` to `mv` function #11145

### PR DESCRIPTION
**Pull request by peter1000**
Adds kw `remove_destination` to `mv` function #11145 incl docs/tests/NEWS entry

### CODE REVIEW

**Comment by peter1000** [comment]
Will fix this soon.

File: base/file.jl [hunk 1]
```diff
@@ -115,7 +115,18 @@ function cp(src::AbstractString, dst::AbstractString; remove_destination::Bool=f
         FS.sendfile(src, dst)
     end
 end
-mv(src::AbstractString, dst::AbstractString) = FS.rename(src, dst)
+
+function mv(src::AbstractString, dst::AbstractString; remove_destination::Bool=false)
+    if ispath(dst)
+        if remove_destination
+            rm(dst; recursive=remove_destination)
```
Comment (tkelman) [inline base/file.jl]: Should this just be `recursive=true` for clarity's sake? It looks like a typo if you glance at it too quickly.

File: base/fs.jl [hunk 2]
```diff
@@ -114,16 +114,12 @@ end
 
 # For move command
 function rename(src::AbstractString, dst::AbstractString)
-    err = ccall(:jl_fs_rename, Int32, (Cstring, Cstring), src, dst)
-
+    err = ccall(:jl_fs_rename, Int32, (Ptr{UInt8}, Ptr{UInt8}), src, dst)
```
Comment (tkelman) [inline base/fs.jl]: why switching this back from Cstring?

File: base/file.jl [hunk 3]
```diff
@@ -115,7 +115,18 @@ function cp(src::AbstractString, dst::AbstractString; remove_destination::Bool=f
...
+        else
+            throw(ArgumentError(string("'$dst' exists. `remove_destination=true` ",
+                                       "is required to remove '$dst' before moving.")))
+        end
+    end
+    rename(src, dst)
```
Comment (tkelman) [inline base/file.jl]: probably still needs to be `FS.rename`

File: base/file.jl (hunk 1 above)
Comment (peter1000) [inline base/file.jl]: it is a typo thanks

**Comment by peter1000** [comment]
any idea why this is marked failed here for **appveyor**

https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.4476 seems to be ok doesn't it?

[appveyor file test](https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.4476/job/urkw0fm8lcaoj26f#L1015)

**Comment by tkelman** [comment]
It said "Appveyor was unable to build non-mergeable pull request" but it may have been a bug. I found the webhook and redelivered it, and it seems okay now.

**Comment by peter1000** [comment]
Thanks a lot.

**Comment by tkelman** [comment]
Any time.

**Comment by peter1000** [comment]
seems this worked out now.

**Comment by tkelman** [comment]
Looks great. Question though (for anyone): what do we want to do for the case when `src == dst`, either exactly, or equivalent in all but case for case-insensitive filesystems, or when one is a link to the other?

**Comment by peter1000** [comment]
> what do we want to do if src == dst, either exactly, or equivalent in all but case for case-insensitive filesystems

I do not have access to Windows but would such case not handled with this PR.

Anyway If you want I could add to the PR some test for such?

>  when one is a link to the other?

It seems to me that the situation is similar to `cp` it was decided that if destination exist it must be purposley [removed first](https://github.com/JuliaLang/julia/blob/master/base/file.jl#L100) and it is up to the coder to check for effects.

```
function cp(src::AbstractString, dst::AbstractString; remove_destination::Bool=false,
                                                         follow_symlinks::Bool=false)
```

Anyway I can think only of one situation where it would be a problem.

`src is a link to dst`  one removes dst and there is nothing left to copy or mv except a broken link.

If you whant I could also add here some more tests.

**Comment by tkelman** [comment]
The problem is `mv(src, src, remove_destination=true)` will delete `src` instead of being a no-op. `mv(src, src)` would fail, but at least not delete anything.

And on case-insensitive filesystems, when `Base.samefile(src, dst)` is true, then `mv(src, dst, remove_destination=true)` will delete `src` instead of changing the case.

**Comment by peter1000** [comment]
> The problem is mv(src, src, remove_destination=true) will delete src instead of being a no-op.

Personally I think if someone call `mv(src, src, remove_destination=true) ` it is there own doing: as they call it purposely with `remove_destination=true` like calling `rm(src)` so I'm unsure this should be a no-op. But we could add it if wanted.

* And on case-insensitive filesystems, when Base.samefile(src, dst) is true, then mv(src, dst, remove_destination=true) will delete dst (and hence also src) instead of changing the case.

If that is the case than it effects also `cp` one idea would be to always first translated the scr/dst to a common case for such case - I think I could do that if wanted

**Comment by tkelman** [comment]
I wasn't suggesting doing any case normalization here. The default file system is case insensitive on Windows and OS X, but can be configured to be case sensitive on both. And I'm not sure we currently have a good way of testing whether or not the filesystem is case sensitive.

But if you call `mv` and are trying to rename a file or directory to a name that only differs in case, the current function will not do what you want at all.

**Comment by tkelman** [comment]
It is a little bit ambiguous what `mv(src, src, remove_destination=true)` should do. But `mv(src, src)` probably shouldn't error.

We might be at the mercy of what libuv is capable of in terms of only changing the case of a file/directory name on a case-insensitive filesystem.

**Comment by peter1000** [comment]
> mv(src, src) probably shouldn't error

`mv(src, src)`  seems to be an unusual special case though and in connection with the requested kwarg (default) `remove_destination=false`  I think it should error. Anyway just how I see it.


the case-insensitive filesystem situation you mentioned makes me a bit more unease.

Anyway: probably still good to merge this for now as it also should fix the other bug you mentioned with sendfile.

**Comment by StefanKarpinski** [comment]
I think that we should make all additional effort to make sure that `mv(src,src)` is handled correctly – i.e. is *always* a no-op, even when remove destination is requested. Likewise, I think we should do whatever it takes to handle `mv(SRC,src)` correctly on all file systems. Usually, this has to be done by moving to a different name and then moving back again. But let's not let it hold up this PR.

**Comment by tkelman** [comment]
> But let's not let it hold up this PR.

I think it should, unless the behavior on master is unequivocally worse AND less safe. The behavior in this PR currently could lead to data loss.

**Comment by peter1000** [comment]
* I can do the things you mentioned about the problem with `(src, src)`

* About the case insensitive systems I do not know how to do this:
> Likewise, I think we should do whatever it takes to handle mv(SRC,src) correctly on all file systems. Usually, this has to be done by moving to a different name and then moving back again. 

I guess the same would apply for `cp` than - do you want to copy it twice?

Anyway I try later to update the PR with the:  `(src, src) ` related additions

**Comment by StefanKarpinski** [comment]
No, the trick would be to do a copy and a rename. Doing this correct and portably is probably hard – which is why it's worth doing in the standard library :-)

**Comment by peter1000** [comment]
> No, the trick would be to do a copy and a rename.

thanks *Stefan* of course you are right I can give this a try.

**Comment by StefanKarpinski** [comment]
Actually, handling this correctly on case-insensitive filesystems isn't really possible and should probably be a no-op or an error: you can't have files named `src` and `SRC` at the same time – they will be the same object.

**Comment by peter1000** [comment]
Reading the comments this is how I see the current situation.

**tkelman**: what do we want to do if src == dst, either exactly, or equivalent in all but case for 
  case-insensitive filesystems, or when one is a link to the other?
            
**StefanKarpinski**: on case-insensitive filesystems isn't really possible and should probably be a 
  no-op or an error: you can't have files named src and SRC at the same time – they will be the same object.
  
Conclusion:

* 1. Seems that the nothing special has to be done for case-insensitive filesystems. 
* 2. Check that `src, dst` are not samefiles which includes also symlinks. raise an error if they refer to the same.

I will update this for all 3 related methods: `cp` helper `cptree` and `mv`

**Comment by ScottPJones** [comment]
@StefanKarpinski Just why can't that be handled correctly on a case-insensitive file system?
Been there, done that... :grinning:
1) Detect that you are both src and dst are actually the same file.  Base.samefile seems to be what is used in Julia.  That's good, because it looks like it is looking at the inodes, so it isn't fooled by hard or symbolic links, or different case being used in either of the paths, or file names (on a case-insensitive file system).
2) If they are the same, check if the file name part only differs by case.
    If they are the same case, this is a no-op.
    If they are different, you must be on a case-insensitive file system, because otherwise you couldn't have ended up with the same inode.
    First rename the file to a temp name (use a UUID, for example).
    Then rename it to a the dst name.

**Comment by tkelman** [comment]
Note that `Base.samefile` currently has some gotchas / unimplemented issues when one or both inputs don't exist, ref #9436

**Comment by peter1000** [comment]
thanks for the hint.
what I do not understand: it says [here](https://github.com/JuliaLang/julia/issues/9436#issuecomment-67921166) that samefile should be exported but it is still not.

**Comment by tkelman** [comment]
That's for a separate PR, none of us have gotten around yet to cleaning it up, adding docs, deciding what to do for 1 or 2 nonexistent inputs, etc.

**Comment by peter1000** [comment]
ah yes I get true for none existing

```
julia> Base.samefile("not_existing1", "not_existing2")
true

julia> 
```

**Comment by peter1000** [comment]
Can I fix at least that check for none existing within `samefile` as I relay on it here and there seems not much point to check it at 2 ends. 

At least it should not return true.

**Comment by tkelman** [comment]
I guess I won't hold it against you if you add a separate commit in here to address the bug part of #9436 and leave the exporting decision (and less clear-cut edge cases) for later.

**Comment by peter1000** [comment]
willl finish tomorrow need to fix the things you pointed out and add tests.

**Comment by peter1000** [comment]
Seems all ok now, isn't it?

File: base/file.jl [hunk 4]
```diff
@@ -71,18 +71,32 @@ end
...
     if ispath(dst)
         if remove_destination
+            # Check for issue when: (src == dst) or when one is a link to the other
+            # https://github.com/JuliaLang/julia/pull/11172#issuecomment-100391076
+            if Base.samefile(src, dst)
+                abs_src = islink(src) ? abspath(readlink(src)) : abspath(src)
+                abs_dst = islink(dst) ? abspath(readlink(dst)) : abspath(dst)
+                throw(ArgumentError(string("'src' and 'dst' refer to the same file/dir.",
+                                           "This is not supported.\n  ",
+                                           "`src` referce to: $(abs_src)\n  ",
```
Comment (tkelman) [inline base/file.jl]: refers

Comment (peter1000) [inline base/file.jl]: Ah sorry will fix that right now.

Comment (tkelman) [inline base/file.jl]: still says "referce," twice

Comment (tkelman) [inline base/file.jl]: looks good now, thanks

**Comment by tkelman** [comment]
There are still some to-do's for fixing things that should be no-ops, or renaming-in-case-only, but for now it looks like those cases will at least error without deleting anything which I guess should be safe enough to merge for now (assuming green CI) and leave to later PR's.

**Pull request merged**